torch/_functorch/_aot_autograd/autograd_cache.py

```diff
@@ -164,8 +222,297 @@ def autograd_cache_hash(
     """
     Generate a unique hash of the FX graph for caching.
     """
+    check_cacheable(gm)
     details = AOTAutogradCacheDetails(gm, example_inputs, config)
     # The prefix distinguishes among the other kinds of objects we cache
     key = "a" + AOTAutogradCachePickler.get_hash(details)
-    log.debug("FX graph cache hash details for key %s:\n%s", key, details.debug_str())
+    log.debug(
+        "Autograd graph cache hash details for key %s:\n%s", key, details.debug_str()
+    )
     return key
+
+
+@dataclass
+class FXGraphCacheLoadable:
+    fx_graph_cache_key: str
+
+    def load(self, example_inputs) -> CompiledFxGraph:
+        # [Note: AOTAutogradCache and FXGraphCache Guard interactions]
+        # FXGraphCache serializes guards that are needed in the shape_env based on the symint inputs to the graph.
+        # he invariant that AOTAutograd uses here is that the sources for symints given to it by dynamo are exactly
+        # the same as the ones it passes to inductor, for both the forward and backward passes.
+        # (This does not mean that the tensor values passed in are the same: only that their symints are).
+        # That is, AOTAutograd and Inductor never create new guards based on symints with different sources
+        # than those passed to it by inductor.
+        result = FxGraphCache._lookup_graph(
+            self.fx_graph_cache_key, example_inputs, True, False
+        )
+        if result is None:
+            log.info("FXGraphCache cache miss for key %s", self.fx_graph_cache_key)
+            raise FXGraphCacheMiss
+        result._boxed_call = True
+        return result
+
+
+@dataclass
+class CompiledForward(FXGraphCacheLoadable):
+    """
+    Cacheable entry for a forward function
+    """
+
+    pass
+
+
+@dataclass
+class CompiledBackward(FXGraphCacheLoadable):
+    """
+    Cacheable entry for a forward function
+    """
+
+    # Used by AOTDispatchAutograd.post_compile
+    backward_state_indices: List[int]
+    num_symints_saved_for_bw_: int
+
+
+@dataclass
+class AOTAutogradCacheEntry:
+    """A single entry into the cache."""
+
+    # Forward and Backward info
+    compiled_fw: CompiledForward
+    compiled_bw: Optional[CompiledBackward]
+
+    # Runtime_metadata saved right before compilation
+    runtime_metadata: ViewAndMutationMeta
+
+    # Wrappers that run after each aot_dispatch_* function
+    dispatch_wrappers: List[CompilerWrapper]
+
+    # Used by AOTSubclassWrapper
+    maybe_subclass_meta: Optional[SubclassMeta]
+    num_fw_outs_saved_for_bw: Optional[int]
+
+    # Used by RuntimeWrapepr
+    indices_of_inps_to_detach: List[int]
+
+    # Turn cache entry into the original callable
+    def wrap_post_compile(
```

Comment:
Maybe worth a comment that this function needs to carefully re-run the exact same set of `post_compile` steps that are run in `aot_dispatch_autograd` and `aot_dispatch_base` during cold start. Just so it's clearer to anyone adding a new runtime calling convention change what code paths need changing.